### ISSUE TITLE
response should be a stream not a json blob

### ISSUE THREAD

**Issue by pmlopes** [vert-x3/vertx-jdbc-client#61]
The current API returns always a single JSON blob with the response from the database. Although this is interesting for simplicity when dealing with a big response data set it will create lot of pressure on the JVM available heap and GC since all rows need to be fetched to RAM, encoded to JSON and sent over the eventbus.

An improvement would be having the concept of cursor where users can fetch one or many rows at once, or a streaming handler.

**Comment by gchauvet** [comment]
I vote for this feature; The current behavior "freeze" my vertx rest service when I have millions rows in my result set (IMO, this is major bottleneck vs node.js ecosystem).

**Comment by vietj** [comment]
can you point out how/where it's done in node.js ? (api)

**Comment by vietj** [comment]
@pmlopes do you mean something like ReadStream<Row>

**Comment by gchauvet** [comment]
We intensively use [brianc/node-pg-query-stream](https://github.com/brianc/node-pg-query-stream) sub-project to manage database query streaming. Adding a similare feature inside this library would be a must have.

**Comment by vietj** [comment]
it seems that `QueryStream` extends Node's `Readable` indeed

**Comment by gchauvet** [comment]
yes, this is the case

**Comment by gchauvet** [comment]
I'm wondering if this feature is to be defined in this library or in vertx-sql-common ?

**Comment by vietj** [comment]
question for @Narigo and @pmlopes

**Comment by pmlopes** [comment]
What I meant when i created the issue was to avoid what @gchauvet is encountering. Currently we copy all data to heap from the DB cursor and then serialize over the eventbus, we probably should handle it as a cursor, read one or a batch of rows, send wait for a request to continue, repeat until no more rows. That is how pg/mongo on node works.

**Comment by pmlopes** [comment]
This however would probably be a huge breaking change, unless we keep the original API

**Comment by pmlopes** [comment]
@gchauvet i do think this should be first designed on sql-common and then implemented both here and in the async client.

**Comment by gchauvet** [comment]
I think this is the best way to proceed, by adding a new signature method like this in SQLConnection  :
```java
SQLConnection query(String sql, Handler<AsyncResult<ReadStream<Row>>> resultHandler);
```

**Comment by vietj** [comment]
not sure ReadStream is yet supported in Handler<AsyncResult<?>> , someone needs to check

**Comment by vietj** [comment]
that being said it could be simply returned, too and not yse Handler<AsyncResult>>, data will be just be pushed

**Comment by gchauvet** [comment]
@vietj :+1: This will also prevent a method clash name caused by genericity.

**Comment by pmlopes** [comment]
due to the codegen strict rules, we can't use the method names:

* query
* update
* etc...

Because overloading is not allowed, even though the signature is completely different.

**Comment by vietj** [comment]
what we cannot do, is having a different return type, overloading is allowed (with some limitations)

(note that having `query(String sql, Handler<AsyncResult<ReadStream<Row>>> resultHandler)` and `query(String sql, Handler<AsyncResult<JsonObject>> resultHandler)` would be forbidden as well but due to Java generic erasure)

Overall it would be good to chose a different name with something like `queryWithStreamingResult` or something like that.

**Comment by vietj** [comment]
note also that it would mean to introduce an API type that extends `ReadStream<Row>` like `StreamingResult` or whatever else (because of lack of reification).

**Comment by gchauvet** [comment]
At the risk of to be considered as an "extremist": why not adopting a full asynchronous API for this library ? The synchronous approach is not very friendly with an event driven architecture ? 

I'm aware this not acceptable before a major version but I think it's worth it.

**Comment by vietj** [comment]
what do you mean by _full asynchronous_ ?

**Comment by gchauvet** [comment]
IMO, full asynchronous mean for me a asynchronous API programming and the use of streams instead collections.

**Comment by pmlopes** [comment]
@gchauvet @vietj @tsegismont @cescoffier would you like to provide some feedback on this proposal?

**Comment by vietj** [comment]
yes

**Comment by vietj** [comment]
is it implemented ?

**Comment by pmlopes** [comment]
@vietj a test impl is available here: vert-x3/vertx-jdbc-client#73
